File: app/src/main/java/com/yoesuv/androidbackgroundservice/prefs/AppPrefHelper.java
package com.yoesuv.androidbackgroundservice.prefs;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.SharedPreferences;
import android.text.TextUtils;

public class AppPrefHelper {

    private static final String DEFAULT_SUFFIX = "_preferences";
    private static final String LENGTH = "#LENGTH";
    private static SharedPreferences mPrefs;

    public static void initPrefs(Context context) {
        new Builder().setContext(context).build();
    }

    private static void initPrefs(Context context, String prefsName, int mode) {
        mPrefs = context.getSharedPreferences(prefsName, mode);
    }

    public static SharedPreferences getPreferences() {
        if (mPrefs != null) {
            return mPrefs;
        }
        throw new RuntimeException(
                "Prefs class not correctly instantiated please call Builder.setContext().build(); in the Application class onCreate.");
    }

    // ===== begin preferences int =====
    public static int getInt(final String key, final int defValue) {
        return getPreferences().getInt(key, defValue);
    }

    public static void putInt(final String key, final int value) {
        final SharedPreferences.Editor editor = getPreferences().edit();
        editor.putInt(key, value);
        editor.apply();
    }
    // ===== end preferences int =====

    public static void remove(final String key) {
        SharedPreferences prefs = getPreferences();
        final SharedPreferences.Editor editor = prefs.edit();
        if (prefs.contains(key + LENGTH)) {
            // Workaround for pre-HC's lack of StringSets
            int stringSetLength = prefs.getInt(key + LENGTH, -1);
            if (stringSetLength >= 0) {
                editor.remove(key + LENGTH);
                for (int i = 0; i < stringSetLength; i++) {
                    editor.remove(key + "[" + i + "]");
                }
            }
        }
        editor.remove(key);

        editor.apply();
    }

    public final static class Builder {

        private String mKey;
        private Context mContext;
        private int mMode = -1;
        private boolean mUseDefault = false;

        public Builder setPrefsName(final String prefsName) {
            mKey = prefsName;
            return this;
        }

        public Builder setContext(final Context context) {
            mContext = context;
            return this;
        }

        public Builder setMode(final int mode) {
            if (mode == ContextWrapper.MODE_PRIVATE || mode == ContextWrapper.MODE_MULTI_PROCESS) {
                mMode = mode;
            } else {
                throw new RuntimeException("The mode in the sharedpreference can only be set too ContextWrapper.MODE_PRIVATE, ContextWrapper.MODE_MULTI_PROCESS");
            }
            return this;
        }

        public Builder setUseDefaultSharedPreference(boolean defaultSharedPreference) {
            mUseDefault = defaultSharedPreference;
            return this;
        }

        public void build() {
            if (mContext == null) {
                throw new RuntimeException("Context not set, please set context before building the Prefs instance.");
            }

            if (TextUtils.isEmpty(mKey)) {
                mKey = mContext.getPackageName();
            }

            if (mUseDefault) {
                mKey += DEFAULT_SUFFIX;
            }

            if (mMode == -1) {
                mMode = ContextWrapper.MODE_PRIVATE;
            }

            AppPrefHelper.initPrefs(mContext, mKey, mMode);
        }

    }

}

File: app/src/main/java/com/yoesuv/androidbackgroundservice/prefs/AppPrefHelper.kt
package com.yoesuv.androidbackgroundservice.prefs

import android.content.Context
import android.content.SharedPreferences
import android.content.ContextWrapper
import android.text.TextUtils
import kotlin.RuntimeException

object AppPrefHelper {

    private const val DEFAULT_SUFFIX = "_preferences"
    private lateinit var mPrefs: SharedPreferences

    private fun initPrefs(context: Context, prefsName: String) {
        mPrefs = context.getSharedPreferences(prefsName, ContextWrapper.MODE_PRIVATE)
    }

    // ===== begin preferences int =====
    fun getInt(key: String?, defValue: Int): Int {
        return mPrefs.getInt(key, defValue)
    }

    fun putInt(key: String?, value: Int) {
        val editor = mPrefs.edit()
        editor.putInt(key, value)
        editor.apply()
    }
    // ===== end preferences int =====

    fun remove(key: String) {
        val editor = mPrefs.edit()
        editor.remove(key)
        editor.apply()
    }

    class Builder {
        private var mKey: String? = null
        private var mContext: Context? = null
        private var mUseDefault = false

        fun setPrefsName(prefsName: String): Builder {
            mKey = prefsName
            return this
        }

        fun setContext(context: Context): Builder {
            mContext = context
            return this
        }

        fun setUseDefaultSharedPreference(defaultSharedPreference: Boolean): Builder {
            mUseDefault = defaultSharedPreference
            return this
        }

        fun build() {
            if (mContext == null) {
                throw RuntimeException("Context not set, please set context before building the Prefs instance.")
            }
            if (TextUtils.isEmpty(mKey)) {
                mKey = mContext?.packageName
            }
            if (mUseDefault) {
                mKey += DEFAULT_SUFFIX
            }

            initPrefs(mContext!!, mKey!!)
        }
    }
}